src/server/project.ts

```diff
@@ -1322,6 +1407,70 @@ namespace ts.server {
         refreshDiagnostics() {
             this.projectService.sendProjectsUpdatedInBackgroundEvent();
         }
+
+        /*@internal*/
+        getPackageJsonsVisibleToFile(fileName: string, rootDir?: string): readonly PackageJsonInfo[] {
+            const packageJsonCache = this.packageJsonCache;
+            const watchPackageJsonFile = this.watchPackageJsonFile.bind(this);
+            const currentDirectory = this.currentDirectory;
+            const getCanonicalFileName = this.getCanonicalFileName;
+            const result: PackageJsonInfo[] = [];
+            const rootPath = rootDir && toPath(rootDir, currentDirectory, this.getCanonicalFileName);
+            forEachAncestorDirectory(getDirectoryPath(fileName), function processDirectory(directory): boolean | undefined {
+                switch (packageJsonCache.directoryHasPackageJson(directory)) {
+                    // Sync and check same directory again
+                    case Ternary.Maybe:
+                        packageJsonCache.searchDirectoryAndAncestors(directory);
+                        return processDirectory(directory);
+                    // Check package.json
+                    case Ternary.True:
+                        const packageJsonFileName = combinePaths(directory, "package.json");
+                        watchPackageJsonFile(packageJsonFileName);
+                        result.push(Debug.assertDefined(packageJsonCache.getInDirectory(directory)));
+                }
+                if (rootPath && rootPath === toPath(directory, currentDirectory, getCanonicalFileName)) {
+                    return true;
+                }
+            });
+
+            return result;
+        }
+
+        /*@internal*/
+        onAddPackageJson(fileName: string) {
+            this.packageJsonCache.addOrUpdate(fileName);
+            this.watchPackageJsonFile(fileName);
+        }
+
+        /*@internal*/
+        getImportSuggestionsCache() {
+            return this.importSuggestionsCache;
+        }
+
+        private watchPackageJsonFile(fileName: string) {
+            const watchers = this.packageJsonFilesMap || (this.packageJsonFilesMap = createMap());
+            if (!watchers.has(fileName)) {
+                watchers.set(fileName, this.projectService.watchFactory.watchFile(

```

Comment:
We dont want to watch file using WatchFile unless its absolutely needed. Eg. for closed script infos, we watch node_modules directory if things happen to fall in node_modules. You would want to do same here to avoid watching using CPU polling on files that more or less dont change without updates to complete directory.

Reply:
As far as I can tell, it’s necessary. I updated the description with a comment explaining:

> @sheetalkamat suggested replacing the additional package.json watch with an existing directory watch that detects changes to node_modules. I tried this approach, but it's quite common to run `npm install --save dep` when `dep` was already in node_modules required by something else, or `npm uninstall --save dep` when `dep` is still required by something else. In these cases, package.json changes, but node_modules does not.

Reply:
IOW, package.json _does_ change without anything else in the directory sometimes, and it would be pretty confusing for auto-imports to be broken in those cases. For better or for worse, people seem to expect a fairly high degree of correctness and completeness out of auto-imports, so I think having one extra watched file on average is worth it.